File: internal/operation/urlreference/protocol/request.go
// Copyright © 2026 @sudocelestia. All rights reserved.
//
// PROPRIETARY AND CONFIDENTIAL SOURCE CODE.
//
// No licence, permission or authorisation is granted to use, copy, modify,
// compile, execute, distribute, publish, sublicense or otherwise exploit this
// file, except to the limited extent unavoidably permitted by applicable law
// or GitHub's Terms of Service.
//
// See the LICENSE file at the repository root for the complete terms.

package workerprotocol

import (
	"bytes"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"strings"
	"time"
	"unicode/utf8"

	"celestia.research/celestia/internal/operation/urlreference/transform"
)

func EncodeRequest(request Request, admittedAt time.Time) ([]byte, error) {
	if err := validateRequest(request, admittedAt); err != nil {
		return nil, err
	}
	data, err := json.Marshal(request)
	if err != nil {
		return nil, protocolError("encode request")
	}
	return data, nil
}

func DecodeRequest(data []byte, admittedAt time.Time) (Request, error) {
	if len(data) == 0 || len(data) > MaxResponseBytes || !utf8.Valid(data) {
		return Request{}, protocolError("request bounds or UTF-8")
	}
	if err := validateJSONFrame(data); err != nil {
		return Request{}, err
	}
	if err := validateRequestFields(data); err != nil {
		return Request{}, err
	}
	var request Request
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&request); err != nil {
		return Request{}, protocolError("decode request")
	}
	if err := validateRequest(request, admittedAt); err != nil {
		return Request{}, err
	}
	return request, nil
}

func validateRequest(request Request, admittedAt time.Time) error {
	if err := validateRequestConstants(request); err != nil {
		return err
	}
	if !validIdentity(request.AttemptID) ||
		!validIdentity(request.RequestNonce) ||
		request.AttemptID == request.RequestNonce {
		return protocolError("request identity")
	}
	if request.Mode != "fang" && request.Mode != "defang" {
		return protocolError("request mode")
	}
	if err := validateRequestInput(request); err != nil {
		return err
	}
	if err := validateDeadline(request.Deadline, admittedAt); err != nil {
		return err
	}
	if request.Limits != (Limits{
		InputBytes:  InputBytes,
		OutputBytes: MaxOutputBytes,
		StderrBytes: StderrBytes,
		MemoryBytes: MemoryBytes,
		Processes:   Processes,
	}) {
		return protocolError("request limits")
	}
	return nil
}

func correlationFor(request Request) responseCorrelation {
	return responseCorrelation{
		attemptID: request.AttemptID,
		nonce:     request.RequestNonce,
		mediaType: request.InputMediaType,
	}
}

func validateRequestConstants(request Request) error {
	if request.ProtocolVersion != ProtocolVersion ||
		request.OperationID != OperationID ||
		request.OperationVersion != OperationVersion ||
		request.InputMediaType != MediaType ||
		request.TimeoutMS != TimeoutMS {
		return protocolError("request constants")
	}
	return nil
}

func validateRequestInput(request Request) error {
	if len(request.Input) < 1 ||
		len(request.Input) > InputBytes ||
		request.InputLength != len(request.Input) {
		return protocolError("request input")
	}
	if err := urlreference.ValidateInput(request.Input); err != nil {
		return protocolError("request URL reference")
	}
	hash := sha256.Sum256([]byte(request.Input))
	if request.InputSHA256 != hex.EncodeToString(hash[:]) {
		return protocolError("request hash")
	}
	return nil
}

func validateDeadline(value string, admittedAt time.Time) error {
	if !strings.HasSuffix(value, "Z") {
		return protocolError("request deadline")
	}
	parsed, err := time.Parse(time.RFC3339Nano, value)
	if err != nil || parsed.Format(time.RFC3339Nano) != value {
		return protocolError("request deadline")
	}
	expected := admittedAt.UTC().Add(
		time.Duration(StartTimeoutMS) * time.Millisecond,
	)
	if !parsed.Equal(expected) {
		return protocolError("request deadline")
	}
	return nil
}

func validIdentity(value string) bool {
	decoded, err := base64.RawURLEncoding.DecodeString(value)
	return err == nil &&
		len(decoded) == 32 &&
		base64.RawURLEncoding.EncodeToString(decoded) == value
}

File: worker/url-reference/src/tests/request.rs
// Copyright © 2026 @sudocelestia. All rights reserved.
//
// PROPRIETARY AND CONFIDENTIAL SOURCE CODE.
//
// No licence, permission or authorisation is granted to use, copy, modify,
// compile, execute, distribute, publish, sublicense or otherwise exploit this
// file, except to the limited extent unavoidably permitted by applicable law
// or GitHub's Terms of Service.
//
// See the LICENSE file at the repository root for the complete terms.

use crate::request::{parse, sha256, valid_deadline, valid_identity, validate};
use crate::transform::transform;

#[test]
fn checks_identifiers() {
    assert!(valid_identity(
        "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8"
    ));
    assert!(!valid_identity(
        "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh9"
    ));
    assert!(!valid_identity("invalid"));
}

#[test]
fn checks_deadlines() {
    assert!(valid_deadline("2026-07-25T12:34:56Z"));
    assert!(valid_deadline("2026-07-25T12:34:56.123456789Z"));
    assert!(valid_deadline("2024-02-29T23:59:59.1Z"));
    assert!(!valid_deadline("2026-07-25T12:34:56+00:00"));
    assert!(!valid_deadline("2026-07-25T12:34:56.Z"));
    assert!(!valid_deadline("2026-02-29T12:34:56Z"));
    assert!(!valid_deadline("2026-13-25T12:34:56Z"));
    assert!(!valid_deadline("2026-07-25T24:34:56Z"));
    assert!(!valid_deadline("2026-07-25T12:34:60Z"));
    assert!(!valid_deadline("2026-07-25T12:34:56.10Z"));
    assert!(!valid_deadline("not-a-deadline"));
}

#[test]
fn mutated_request_frames_are_deterministic_and_bounded() {
    let seed = valid_request_frame();
    assert!(parse(&seed).is_ok());
    let mut state = 0x9e37_79b9_7f4a_7c15_u64;
    for iteration in 0..25_000 {
        state ^= state << 13;
        state ^= state >> 7;
        state ^= state << 17;
        let mut candidate = seed.clone();
        let index =
            usize::try_from(state).expect("u64 fits usize on supported workers") % candidate.len();
        match iteration % 4 {
            0 => candidate[index] ^= u8::try_from(state >> 56).expect("shifted byte fits u8"),
            1 => {
                candidate.remove(index);
            }
            2 => candidate.insert(
                index,
                u8::try_from(state >> 56).expect("shifted byte fits u8"),
            ),
            _ => candidate.truncate(index),
        }
        let first = parse(&candidate);
        let second = parse(&candidate);
        assert_eq!(first.is_ok(), second.is_ok());
        if let Ok(request) = first {
            assert!(validate(&request).is_ok());
            assert!(transform(&request.input, &request.mode).is_ok());
        }
    }
}

fn valid_request_frame() -> Vec<u8> {
    let input = "https://example.test/";
    format!(
        concat!(
            "{{\"protocol_version\":1,\"operation_id\":\"url-reference\",",
            "\"operation_version\":1,",
            "\"attempt_id\":\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8\",",
            "\"request_nonce\":\"AQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyA\",",
            "\"input_media_type\":\"text/plain; charset=utf-8\",",
            "\"input_length\":{},\"input_sha256\":\"{}\",",
            "\"mode\":\"defang\",\"deadline\":\"2026-07-25T12:34:56Z\",",
            "\"timeout_ms\":2000,\"limits\":{{\"input_bytes\":4096,",
            "\"output_bytes\":8192,\"stderr_bytes\":8192,",
            "\"memory_bytes\":67108864,\"processes\":1}},",
            "\"input\":\"{}\"}}"
        ),
        input.len(),
        sha256(input),
        input,
    )
    .into_bytes()
}
